spark/src/main/java/org/apache/iceberg/spark/actions/BaseRemoveFilesSparkAction.java

```diff
@@ -0,0 +1,220 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.actions;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+import org.apache.iceberg.HasTableOperations;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableMetadataParser;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.actions.RemoveFiles;
+import org.apache.iceberg.actions.RemoveFilesActionResult;
+import org.apache.iceberg.exceptions.NotFoundException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.spark.JobGroupInfo;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.Tasks;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Encoders;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An action that performs the same operation as {@link RemoveFiles} but uses Spark
+ * to determine the files that needs to be deleted. The action uses metadata tables to
+ * find the files to be deleted.
+ * Deletes are performed locally after retrieving the results from the Spark executors.
+ */
+@SuppressWarnings("UnnecessaryAnonymousClass")
+public class BaseRemoveFilesSparkAction
+    extends BaseSparkAction<RemoveFiles, RemoveFiles.Result> implements RemoveFiles {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseRemoveFilesSparkAction.class);
+
+  private static final String DATA_FILE = "Data File";
+  private static final String MANIFEST = "Manifest";
+  private static final String MANIFEST_LIST = "Manifest List";
+  private static final String OTHERS = "Others";
+
+  private static final String STREAM_RESULTS = "stream-results";
+
+  // Creates an executor service that runs each task in the thread that invokes execute/submit.
+  private static final ExecutorService DEFAULT_DELETE_EXECUTOR_SERVICE = null;
+
+  private final Table table;
+  private final TableOperations ops;
+  private final Consumer<String> defaultDelete = new Consumer<String>() {
+    @Override
+    public void accept(String file) {
+      ops.io().deleteFile(file);
+    }
+  };
+
+  private Consumer<String> deleteFunc = defaultDelete;
+  private ExecutorService deleteExecutorService = DEFAULT_DELETE_EXECUTOR_SERVICE;
+
+  public BaseRemoveFilesSparkAction(SparkSession spark, Table table) {
```

Comment:
The main purpose of this action is to remove files once a table dropped. How do we see that happening? Will we try to load the table before drop, then drop the table, then call this action to cleanup? That way we get the correct fileIO?

cc @RussellSpitzer @rdblue too

Reply:
We could accept a metadata file location as an alternative.

```
actions.removeFiles(jsonFileLocation)
    .fileIO(io)
    .execute();
```

Reply:
I think the second option sends a message the table has been dropped and we just point to a version file. Plus, this will enable cleaning up files not only during drop but also later. Users won't have to about about `StaticTableOperations`.

Reply:
I like passing the JSON file location, but we would probably want to make it clear that it deletes the whole tree referenced by the metadata file. I don't think that's clear from just the code sample above.

Reply:
@rdblue, any ideas on naming or other ways to convey that?

`removeFileTree(jsonFileLocation)` or something?

Reply:
I consider this action as something we would mostly use from Iceberg code. For example, we currently drop the table and then call `CatalogUtil$dropTableData` that will try to clean as much as possible but may also fail at some point. I see this as a replacement of that logic that should scale much better as we will use metadata tables.

I hope users will not really interact with this action and just use the regular DROP with or without purge. The only case when someone would need to invoke this manually is if something went wrong during DROP. In that case, it seems reasonable to ask the user to check whether the table was dropped and find a recent metadata file to use for cleaning the file tree. By calling this as `RemoveReachableFiles` and accepting a metadata location instead of a table we send a message it is something that is invoked after DROP.

Reply:
Oh I see, yea if we can use this code as part of drop, that would be great.  

I did hit issues where drop fails, and leaves behind files but the table is already gone.  In this case, deleting the metadata json last makes good sense for recovery, thanks for the context, I suppose we can add this documentation of the action if not done already.